repo/mk/js.deps.npm.mk

```diff
@@ -79,3 +76,22 @@ deps-npm-prod:
 		$(NPM) update --no-save --production; \
 	}
 	$(NPM) list --depth=0 || $(MAKE) deps-npm-unmet-peer
+
+
+# the audit can be ignored by adding to the Makefile:
+# SF_DEPS_TARGETS := $(filter-out deps-npm-audit,$(SF_DEPS_TARGETS))
+
+.PHONY: deps-npm-audit
+deps-npm-audit:
+	$(ECHO_DO) "Auditing package.json..."
+	$(GIT) ls-files --error-unmatch "package-lock.json" || {
+		$(RM) package-lock.json
+		$(NPM) install --package-lock-only
+	}
+	$(NPM) audit || { \
+		$(NPM) audit fix; \
+		$(GIT) ls-files --error-unmatch "package-lock.json" || $(RM) package-lock.json; \
+		exit 1; \
```

Comment:
Do we need to exit with error here? As I understand it, if `npm audit fix` succeeds, we should happily continue, right? Or no?

Reply:
it is actually the same pattern as we use for the linters, see https://github.com/tobiipro/support-firecloud/blob/master/repo/mk/js.check.eslint.mk#L28-L30

it fails, because at least the CI should fail on this one, as the audit fixes should be checked in.

there are 3 situations to handle with `npm audit` failing, all requiring a commit:

1. `package-lock.json` is checked-in, then after `npm audit fix` would modify it and maybe `package.json` too, and you should check them in
2. `package-lock.json` is NOT checked-in, `npm audit fix` might modify `package.json`, and you should check it in
3. `package-lock.json` is NOT checked-in, `npm audit fix` cannot fix the security by upgrading some top-level dependency in `package.json`, and basically you have 2 options, either don't run deps-npm-audit, or commit a `package-lock.json` (smth which I don't mention on purpose because). If you don't do any of these 2 things, then there is no way to get passed this check.

Reply:
I agree with these points. More worried about CI and development experience on branches. It can be so that a feature branch is suddenly starting to fail without no apparent reason and then it has to include changes that are not related to the feature. 
My suggestion is to run audit only on master. So we have such cases:
1. Master - has audits, runs on cron and fails early, always gets fixed ASAP by someone from the team.
2. Feature branches - don't run audit, it will be run on master after merge.
3. Provision branches (env/X) - don't run the audit, install whatever dependencies specified in a given tag. We make deploy more reliable. 

WDYT?

Reply:
> It can be so that a feature branch is suddenly starting to fail without no apparent reason and then it has to include changes that are not related to the feature.

That is the case for plenty of other things. That is not to say that +1 doesn't matter, but that it's a risk that is taken as a development feature - fail early.

> My suggestion is to run audit only on master.

There a few issues with that.
**First** - a practicality. The Makefiles in support-firecloud should not have such exceptions, so that leaves having an ìf` in each project's Makefile to `filter-out` (or if we decide this is an opt-in target, add) the `deps-npm-audit` target.
**Secondly** - it opens the door for surprises after merging the PR. You add packages for your feature, get the PR accepted, and then master fails. The purpose of PRs is to keep master as clean as possible for a variety of reasons, but we've just hijacked that feature now. And on top of it, maybe it is not possible to make a fix so that `npm audit` is green, maybe you need to exchange a whole package with another, etc, etc maybe it takes time, so you need to revert your PR merge, etc

> Provision branches (env/X) - don't run the audit, install whatever dependencies specified in a given tag. We make deploy more reliable.

I don't think we run this there. We take the snapshots. Or?

Reply:
> That is the case for plenty of other things. 

And all of them are annoying and disruptive when you have to figure out why the hell your branch starts to fail because someone pushed some release in a dependency :) I wouldn't want there to be one more.

> First - a practicality. The Makefiles in support-firecloud should not have such exceptions, so that leaves having an if in each project's Makefile to filter-out (or if we decide this is an opt-in target, add) the deps-npm-audit target.

Another way: put it in the `ci` script and create a top level target `make audit` that will be run only on master. There is a number of ways to do it and you know it much better.

> Secondly - it opens the door for surprises after merging the PR. You add packages for your feature, get the PR accepted, and then master fails. The purpose of PRs is to keep master as clean as possible for a variety of reasons, but we've just hijacked that feature now. And on top of it, maybe it is not possible to make a fix so that npm audit is green, maybe you need to exchange a whole package with another, etc, etc maybe it takes time, so you need to revert your PR merge, etc

That's not such a surprise, actually, since Github runs PR checks against a potential merge commit with the master, so we will fail before the actual merge. Another option: to do the audit on `make deps` as we do now without failure. I pay attention to those warnings and would be even more attentive if I knew that it will break master after merging in my branch.

> I don't think we run this there. We take the snapshots. Or?

You are probably right 👍